snap/integrity/dmverity/veritysetup.go

```
@@ -122,33 +124,127 @@ func shouldApplyNewFileWorkaroundForOlderThan204() (bool, error) {
+		hashDevice,
+	}
+
+	if opts != nil {
+		args = opts.appendArguments(args)
+	}
+
+	output, stderr, err := osutil.RunSplitOutput("veritysetup", args...)
 	if err != nil {
-		return nil, osutil.OutputErrCombine(output, stderr, err)
+		return "", osutil.OutputErrCombine(output, stderr, err)
 	}
 
-	logger.Debugf("cmd: 'veritysetup format %s %s':\n%s", dataDevice, hashDevice, string(output))
+	logger.Debugf("cmd: 'veritysetup format %s %s %s':\n%s", dataDevice, hashDevice, args, string(output))
 
 	rootHash, err := getRootHashFromOutput(output)
+	if err != nil {
+		return "", err
+	}
+
+	return rootHash, nil
+}
+
+// VeritySuperblock represents the dm-verity superblock structure.
+//
+// It mirrors cryptsetup's verity_sb structure from
+// https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/verity/verity.c?ref_type=heads#L25
+type VeritySuperBlock struct {
+	Signature       [8]uint8   `json:"-"`               /* "verity\0\0" */
+	Version         uint32     `json:"version"`         /* superblock version */
+	Hash_type       uint32     `json:"hash_type"`       /* 0 - Chrome OS, 1 - normal */
+	Uuid            [16]uint8  `json:"uuid"`            /* UUID of hash device */
+	Algorithm       [32]uint8  `json:"algorithm"`       /* hash algorithm name */
+	Data_block_size uint32     `json:"data_block_size"` /* data block in bytes */
+	Hash_block_size uint32     `json:"hash_block_size"` /* hash block in bytes */
+	Data_blocks     uint64     `json:"data_blocks"`     /* number of data blocks */
+	Salt_size       uint16     `json:"salt_size"`       /* salt size */
+	Pad1            [6]uint8   `json:"-"`
+	Salt            [256]uint8 `json:"salt"` /* salt */
+	Pad2            [168]uint8 `json:"-"`
+}
+
+func (sb *VeritySuperBlock) Size() int {
+	size := int(unsafe.Sizeof(*sb))
+	return size
+}
+
+// Validate will perform consistency checks over an extracted superblock to determine whether it's a valid
+// superblock or not.
+func (sb *VeritySuperBlock) Validate() error {
```

Should do this more checks, for instance look at whether Algorithm has a valid value, etc.?

---

My intention with this was to simply do some rudimentary checks to fail early. Also check how I am using it in the [subsequent PR](https://github.com/canonical/snapd/pull/14872/commits/a646fa879f7fc6c148b8a7fafb20ff53200a4954#diff-f2b95ee6cb86d0d94e54d356707f39c2e6014beef108da27d124daf04d6eb5daR61). After calling `Validate()` I do a more specific check that the algorithm found in the superblock is matching the one we want to generate.

In a similar spirit to the comment in the other PR, what should we consider as a valid value here? Any algorithm supported by dm-verity or be specific and statically check that it is `sha256`, which is the only algorithm we want to support for now? I will give this some more thought.

---

Also see https://github.com/canonical/snapd/pull/14871#discussion_r1914771307 where I explain what I intend to use the superblock for.